multi_partner_learning.py

```diff
@@ -302,6 +315,112 @@ def compute_collaborative_round_fedavg(self):
 
         logger.debug("End of fedavg collaborative round.")
 
+
+    def compute_collaborative_round_qavg(self):
+            """Proceed to a collaborative round with a quality aware federated averaging approach"""
+
+            logger.debug("Start new qavg collaborative round ...")
+
+            # Initialize variables
+            epoch_index, minibatch_index = self.epoch_index, self.minibatch_index
+            is_very_first_minibatch = (epoch_index == 0 and minibatch_index == 0)
+            x_val, y_val = self.val_data
+            partners_list = self.partners_list
+
+            # Starting model for each partner is the aggregated model from the previous mini-batch iteration
+            if is_very_first_minibatch:  # Except for the very first mini-batch where it is a new model
+                logger.debug(f"(fedavg) Very first minibatch of epoch n°{epoch_index}, init new models for each partner")
+                partners_model_list_for_iteration = self.init_with_new_models()
+            else:
+                logger.debug(f"(fedavg) Minibatch n°{minibatch_index} of epoch n°{epoch_index}, "
+                            f"init aggregated model for each partner with models from previous round")
+                partners_model_list_for_iteration = self.init_with_agg_models()
+
+            # model_to_evaluate = partners_model_list_for_iteration[0]
+            # model_evaluation = model_to_evaluate.evaluate(x_val, y_val, batch_size=constants.DEFAULT_BATCH_SIZE, verbose=0)
+            # self.score_matrix_collective_models[epoch_index, minibatch_index] = model_evaluation[1]
+            
+
+            # Select partners for model testing
+            # We use a method of divergence between two partners based on label distribution difference between them.
+            # We select k points which are closest from all label repartition vectors barycenter
+            labels = list(set(y_val))
+            
+            mean_label_repartion = {label:0 for label in labels}
+
+            for p in partners_list:
+                
+                for label,size in zip(p.label_repartition):
+                    
+                    mean_label_repartion[label] += size
+
+            for label,size in zip(mean_label_repartion):
+
+                mean_label_repartion[label]/= len(partners_list)
```

Comment:
OK so if I understand correctly this is based on a distance in terms of number of samples per class, right?